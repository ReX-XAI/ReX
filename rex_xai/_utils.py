#!/usr/bin/env python3

import importlib.metadata
from itertools import chain, combinations
from enum import Enum
from typing import Tuple, Union, Dict
from numpy.typing import NDArray
import torch as tt
import numpy as np
from skimage.segmentation import mark_boundaries
from rex_xai.logger import logger
from rex_xai.box import Box

Strategy = Enum("Strategy", ["Global", "Spatial", "MultiSpotlight", "Contrastive"])

Queue = Enum("Queue", ["Area", "All", "Intersection", "DC"])

SpatialSearch = Enum("SpatialSearch", ["NotFound", "Found"])


def powerset(r, reverse=True):
    ps = list(chain.from_iterable(combinations(r, lim) for lim in range(1, len(r) + 1)))
    if reverse:
        return reversed(ps)
    else:
        return ps


def clause_area(clause, areas: Dict) -> int:
    tot = 0
    for c in clause:
        tot += areas[c]
    return tot


class ReXError(Exception):
    pass


class ReXTomlError(ReXError):
    def __init__(self, message) -> None:
        self.message = message
        super().__init__(self.message)

    def __str__(self) -> str:
        return f"ReXTomlError: {self.message}"


class ReXPathError(ReXError):
    def __init__(self, message) -> None:
        self.message = message
        super().__init__(self.message)

    def __str__(self) -> str:
        return f"ReXPathError: no such file exists at {self.message}"


class ReXScriptError(ReXError):
    def __init__(self, message) -> None:
        self.message = message
        super().__init__(self.message)

    def __str__(self) -> str:
        return f"ReXScriptError: {self.message}"


class ReXDataError(ReXError):
    def __init__(self, message) -> None:
        self.message = message
        super().__init__(self.message)

    def __str__(self) -> str:
        return f"ReXDataError: {self.message}"


class ReXMapError(ReXError):
    def __init__(self, message) -> None:
        self.message = message
        super().__init__(self.message)

    def __str__(self) -> str:
        return f"ReXMapError: {self.message}"


def xlogx(ps):
    f = np.vectorize(_xlogx)
    return f(ps)


def _xlogx(p):
    if p == 0.0:
        return 0.0
    else:
        return p * np.log2(p)


def add_boundaries(
    img: Union[NDArray, tt.Tensor], segs: NDArray, colour=None
) -> NDArray:
    if colour is None:
        m = mark_boundaries(img, segs, mode="thick")
    else:
        m = mark_boundaries(img, segs, colour, mode="thick")
    m *= 255  # type: ignore
    m = m.astype(np.uint8)
    return m


def get_device(gpu: bool):
    if tt.backends.mps.is_available() and gpu:
        return tt.device("mps")
    if tt.device("cuda") and gpu:
        return tt.device("cuda")
    if gpu:
        logger.warning("gpu not available")
    return tt.device("cpu")


def get_map_locations(map, reverse=True):
    if isinstance(map, tt.Tensor):
        map = map.detach().cpu().numpy()
    coords = []
    for i, r in enumerate(np.nditer(map)):
        coords.append((r, np.unravel_index(i, map.shape)))
    coords = sorted(coords, reverse=reverse)
    return coords


def set_boolean_mask_value(
    tensor,
    mode,
    order,
    coords: Union[Box, Tuple[NDArray, NDArray]],
    val: bool = True,
):
    if isinstance(coords, Box):
        if mode in ("spectral", "tabular"):
            h = coords.col_start
            w = coords.col_stop
        else:
            h = slice(coords.row_start, coords.row_stop)
            w = slice(coords.col_start, coords.col_stop)
    else:
        h = coords[0]
        w = coords[1]
    # three channels
    if mode == "RGB":
        # (C, H, W)
        if order == "first":
            tensor[:, h, w] = val
        # (H, W, C)
        else:
            tensor[h, w, :] = val
    elif mode == "L":
        if order == "first":
            # (1, H, W)
            tensor[0, h, w] = val
        else:
            tensor[h, w, :] = val
    elif mode in ("spectral", "tabular"):
        if len(tensor.shape) == 1:
            tensor[h:w] = val
        else:
            tensor[0, h:w] = val
    # elif mode == "tabular":

    elif mode == "voxel":
        logger.warning("not yet implemented")


def ff(obj, fmt):
    """
    Like format(obj, fmt), but returns the string 'None' if obj is None.
    See the help for format() to see acceptable values for fmt.
    """
    return "None" if obj is None else format(obj, fmt)


def version():
    return importlib.metadata.version("rex-xai")
