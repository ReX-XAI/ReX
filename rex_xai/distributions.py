"""distributions module"""

from typing import Optional, Tuple
from enum import Enum
import numpy as np
from scipy.stats import binom, betabinom
from rex_xai.logger import logger

Distribution = Enum("Distribution", ["Binomial", "Uniform", "BetaBinomial", "Adaptive"])


def _betabinom2d(height, width, alpha, beta):
    bx = betabinom(width, alpha, beta)
    by = betabinom(height, alpha, beta)

    w = np.array([bx.pmf(i) for i in range(0, width)])  # type: ignore
    h = np.array([by.pmf(i) for i in range(0, height)])  # type: ignore

    w = np.expand_dims(w, axis=0)
    h = np.expand_dims(h, axis=0)

    u = (h.T * w / np.sum(h.T * w)).ravel()
    p = np.random.choice(np.arange(0, len(u)), p=u)
    return p


def _blend(dist, alpha, base):
    pmf = np.array([base.pmf(x) for x in range(0, len(dist))])
    blend = ((1.0 - alpha) * pmf) + (alpha * dist)
    blend /= np.sum(blend)
    return blend


def _2d_adaptive(map, args: Tuple[int, int, int, int], alpha=0.0, base=None) -> int:
    # if the map exists and is not 0.0 everywhere...
    if map is not None and np.max(map) > 0.0:
        s = map[args[0] : args[1], args[2] : args[3]]
        sf = np.ndarray.flatten(s)
        # sf = np.max(sf) - sf
        sf /= np.sum(sf)

        # base = betabinom(0, len(sf), 1.1, 1.1)
        # if base is not None:
        #     sf = _blend(alpha, base)
        pos = np.random.choice(np.arange(0, len(sf)), p=sf)
        return pos

    # if the map is empty or doesn't exist, return uniform
    return np.random.randint(1, (args[1] - args[0]) * (args[3] - args[2]))


def str2distribution(d: str) -> Distribution:
    """converts string into Distribution enum"""
    if d == "binom":
        return Distribution.Binomial
    elif d == "uniform":
        return Distribution.Uniform
    elif d == "betabinom":
        return Distribution.BetaBinomial
    elif d == "adaptive":
        return Distribution.Adaptive
    else:
        logger.warning(
            "Invalid distribution '%s', reverting to default value Distribution.Uniform",
            d,
        )
    return Distribution.Uniform


def random_coords(d: Optional[Distribution], *args, map=None) -> Optional[int]:
    """generates random coordinates given a distribution and args"""

    try:
        if d == Distribution.Adaptive:
            return _2d_adaptive(map, args[0])

        if d == Distribution.Uniform:
            return np.random.randint(1, args[0])  # type: ignore

        if d == Distribution.Binomial:
            start, stop, *dist_args = args[0]
            return binom(stop - start - 1, dist_args).rvs() + start

        if d == Distribution.BetaBinomial:
            return _betabinom2d(args[1], args[2], args[3][0], args[3][1])

        return None
    except ValueError:
        return None
