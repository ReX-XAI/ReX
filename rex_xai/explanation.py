#!/usr/bin/env python3
# from __future__ import annotations
"""main logical entrypoint for ReX."""

import copy
import os
import sys
import time
from typing import Tuple, List, Union

from scipy.io import loadmat
import numpy as np
import torch as tt
from PIL import Image
from sqlalchemy.orm import Session
from tqdm import trange  # type: ignore

from rex_xai.config import CausalArgs
from rex_xai.database import update_database
from rex_xai.evaluation import Evaluation
from rex_xai.multi_explanation import MultiExplanation
from rex_xai.extraction import Explanation
from rex_xai.input_data import Data
from rex_xai.logger import logger
from rex_xai.onnx import get_prediction_function
from rex_xai.resp_maps import ResponsibilityMaps
from rex_xai.responsibility import causal_explanation
from rex_xai.prediction import Prediction
from rex_xai._utils import Strategy, ReXScriptError


def try_preprocess(args: CausalArgs, model_shape: Tuple[int], device: tt.device):
    """Makes an attempt to preprocess input data as required for the model.

    Data preprocessing is based on file extension and (possibly) user-defined mode.
    File extensions in ``[".jpg", ".jpeg", ".png", ".tif", ".tiff"]`` are treated
    as images, ".npy" and ".mat" are treated as Numpy arrays, and ".nii" are treated as nifti files.
    For any other file extension, we create a ``Data`` object without pre-processing.

    Args:
        args: configuration values for ReX
        model_shape: shape of the input tensor of the model, as returned by
            :py:func:`~rex_xai.explanation.get_prediction_func_from_args()`
        device: as returned by :py:func:`~rex_xai._utils.get_device()`

    Returns:
      Data: the processed input data
    """
    _, ext = os.path.splitext(args.path)
    # args.path is an image
    if ext.lower() in [".jpg", ".jpeg", ".png", ".tif", ".tiff"]:
        # a simple sanity check: spectral_occlusion is not suitable for images
        # so remind the user to update rex.toml and set mask_value to 0 in the meantime
        try:
            if args.mask_value in ("tabular", "spectral"):  # type: ignore
                logger.warning(
                    f"{args.mask_value} is not suitable for images. Changing mask_value to 0"
                )
                args.mask_value = 0
        except AttributeError:
            pass

        img = Image.open(args.path)
        if img.mode == "RGBA":
            logger.warning("RGBA input image provided, converting to RGB")
            img = img.convert("RGB")

        data = Data(img, model_shape, device)
        data.generic_image_preprocess(means=args.means, stds=args.stds, norm=args.norm)

    # a numpy "npy" array or matlab "mat" file
    elif ext in (".npy", ".mat"):
        if args.mode in ("tabular", "spectral"):
            if ext == ".mat":
                raw_data = np.load(loadmat(args.path)["val"])
            else:
                raw_data = np.load(args.path)
            data = Data(raw_data, model_shape, mode=args.mode, device=device)
            data.data = data.generic_tab_preprocess()
        else:
            logger.fatal("we do not generically handle this datatype")
            return NotImplemented
    # nifti files for 3d data
    elif ext in ".nii":
        logger.fatal("we do not (yet) handle nifti files generically")
        return NotImplemented
    else:
        # we don't know what to do!
        data = Data(args.path, model_shape, mode=args.mode, device=device)

    return data


def load_and_preprocess_data(
    model_shape: Tuple[int], device: tt.device, args: CausalArgs
):
    """Loads input data from filepath and does preprocessing.

    Uses a custom preprocesssing function if this is defined in ``args.script.preprocess``,
    otherwise :py:func:`~rex_xai.explanation.try_preprocess()`.

    Args:
        model_shape: shape of the input tensor of the model, as returned by
            :py:func:`~rex_xai.explanation.get_prediction_func_from_args()`
        device: as returned by :py:func:`~rex_xai._utils.get_device()`
        args: configuration values for ReX

    Returns:
        Data: the processed input data

    """
    if args.script is not None:
        if hasattr(args.script, "preprocess"):
            data = args.script.preprocess(
                args.path, model_shape, device, mode=args.mode
            )
        else:
            raise ReXScriptError(
                f"{args.script_location} is missing a preprocess() function"
            )
    else:
        # no custom preprocessing, so we make our best guess as to what to do
        data = try_preprocess(args, model_shape, device)

    return data


def predict_target(data: Data, prediction_func) -> Prediction:
    """Predicts classification of input data, using given prediction function.

    Uses ``prediction_func`` to identify the classification of the input data and return
    this as the target classification for ReX.

    Args:
        data: processed input data object
        args: configuration values for ReX
        prediction_func: prediction function for the model

    Returns:
        Prediction: the predicted target classification and confidence
    """
    target = prediction_func(data.data, None)

    if isinstance(target, list):
        target = target[0]

    if target is not None:
        logger.info(
            "image classified as %s with %f confidence",
            target.classification,
            target.confidence,
        )
    else:
        logger.warning("no target found")
        sys.exit(-1)

    return target


def calculate_responsibility(
    data: Data, args: CausalArgs, prediction_func, keep_all_maps=False
) -> tuple[ResponsibilityMaps, dict]:
    """Calculates ResponsibilityMaps for input data using given args.

    Runs :py:func:`~rex_xai.responsibility.causal_explanation` for ``args.iters`` iterations,
    and returns a ResponsibilityMaps object and a dictionary containing some statistics about
    the calculation process.
    The ResponsibilityMaps object by default only includes the responsibility map that matches
    the classification of the input data. Set ``keep_all_maps`` to ``True`` to retain all maps.

    Args:
        data: processed input data object
        args: configuration values for ReX
        prediction_func: prediction function for the model
        keep_all_maps: whether to retain all :py:class:`~rex_xai.resp_maps.ResponsibilityMaps`,
            or just the one that matches the target classification.

    Returns:
        tuple containing

        - ResponsibilityMaps: ResponsibilityMaps for the given data, prediction function, and args.
        - dict: statistics for the call of this function that generated the ResponsibilityMaps object
    """

    if data.target is None or data.target.classification is None:
        raise ValueError(
            "No target classification found. Please run `predict_target` before running `calculate_responsibility`."
        )

    maps = ResponsibilityMaps()
    if data.model_height is not None:
        maps.new_map(
            data.target.classification,
            data.model_height,
            data.model_width,
            data.model_depth,
        )
    else:
        maps.new_map(data.target.classification, data.model_height, data.model_width)

    total_passing: int = 0
    total_failing: int = 0
    max_depth_reached: int = 0
    avg_box_size: float = 0.0

    if args.iters > 0:
        for i in trange(args.iters, disable=not args.progress_bar):
            (
                local_maps,
                passing,
                failing,
                depth_reached,
                avg_bs,  # average box size
            ) = causal_explanation(
                i + 1,
                data,
                args,
                prediction_func,
                current_map=maps.get(data.target.classification),
            )

            total_passing += passing
            total_failing += failing
            max_depth_reached = max(max_depth_reached, depth_reached)
            avg_box_size += avg_bs
            # TODO this needs to be smarter. If we only have shallow penetration, then
            # this is doing us a disservice. Perhaps leave merging until after completion
            # of all iterations. Might potentially use a lot of memeory though
            if depth_reached > 1:
                maps.merge(local_maps)

    avg_box_size /= args.iters

    logger.info(
        "Total Passing: %d, Total Failing %d, Max Depth Reached %d, Avg Box Size %f",
        total_passing,
        total_failing,
        max_depth_reached,
        avg_box_size,
    )

    run_stats = {
        "total_passing": total_passing,
        "total_failing": total_failing,
        "max_depth_reached": max_depth_reached,
        "avg_box_size": avg_box_size,
    }

    return maps, run_stats


def analyze(exp: Explanation, data_mode: str | None):
    """Analyzes an Explanation.

    Analyzes the area ratio, entropy difference, insertion and deletion curves for an
    Explanation object, prints them, and returns them.

    Args:
        exp: Explanation object as returned by :py:func:`~rex_xai.explanation._explanation`
        data_mode: Mode of the input data. Entropy difference is only calculated if ``data_mode``
            is one of ["RGB", "L"].

    Returns:
        tuple containing

        - area (float)
        - entropy_diff (float)
        - insertion_curve (float)
        - deletion_curve (float)

    """
    eval = Evaluation(exp)
    rat = eval.ratio()
    ent = None
    max_ent = None
    if data_mode in ("RGB", "RGBA", "L"):
        be, ae = eval.entropy_loss()  # type: ignore
        ent = be - ae
    elif data_mode in ("spectral", "tabular"):
        ent, max_ent = eval.spectral_entropy()

    iauc, dauc = eval.insertion_deletion_curve(
        exp.prediction_func, normalise=exp.args.normalise_curves
    )

    analysis_results = {
        "area": rat,
        "entropy": ent,
        "max_entropy": max_ent,
        "insertion_curve": iauc,
        "deletion_curve": dauc,
    }

    return analysis_results


def _explanation(
    args: CausalArgs,
    model_shape: Tuple[int],
    prediction_func,
    device: tt.device,
    db: Session | None = None,
    path=None,
):
    """Takes a CausalArgs object and model information and returns a Explanation.

    Takes a CausalArgs object, model shape and prediction function and returns an Explanation.
    Depending on the input ``args``, optionally produces output plots, analyses the output
    explanation, and/or writes results to a database.

    Args:
        args: configuration values for ReX
        model_shape: shape of the input tensor of the model, as returned by :py:func:`~rex_xai.explanation.get_prediction_func_from_args()`
        prediction_func: as returned by :py:func:`~rex_xai.explanation.get_prediction_func_from_args()`
        device: as returned by :py:func:`~rex_xai._utils.get_device()`
        db: None or as returned by :py:func:`~rex_xai.database.initialise_rex_db()`

    Returns:
        Explanation:
            An :py:class:`~rex_xai.extraction.Explanation` object containing the causal reponsibility explanation
            calculated using the given ``args``.

    """
    data = load_and_preprocess_data(model_shape, device, args)
    data.set_mask_value(args.mask_value)
    logger.debug(
        "args.mask_value is %s, data.mask_value is %s", args.mask_value, data.mask_value
    )

    data.target = predict_target(data, prediction_func)

    start = time.time()

    logger.info("Calculating responsibility map")
    resp_object, run_stats = calculate_responsibility(data, args, prediction_func)

    logger.info("Extracting explanation from responsibility map")
    clauses = None
    if args.strategy in (Strategy.MultiSpotlight, Strategy.Contrastive):
        exp = MultiExplanation(resp_object, prediction_func, data, args, run_stats)
        exp.extract()

        if args.strategy == Strategy.Contrastive and args.permitted_overlap != 1.0:
            logger.warning(
                "contrastive explanations require a permitted overlap of 1.0, so setting this now"
            )
            args.permitted_overlap = 1.0

        clauses = exp.separate_by(args.permitted_overlap)
        logger.info(f"found the following sets of explanations {clauses}")

        if args.strategy == Strategy.Contrastive:
            clauses = exp.contrastive(clauses)
            args.multi_style = "contrastive"
    else:
        exp = Explanation(resp_object, prediction_func, data, args, run_stats)
        exp.extract(args.strategy)

    if args.analyze:
        logger.info("Analysing explanation")
        results = analyze(exp, data.mode)
        if data.mode == "spectral":
            print(
                f"INFO:ReX:classification {exp.data.target.classification}, area {results['area']}, responsibility entropy {results['entropy']},",  # type: ignore
                f"max entropy {results['max_entropy']}",
                f"insertion curve {results['insertion_curve']}, deletion curve {results['deletion_curve']}",
            )
        else:
            print(
                f"INFO:ReX:classification {exp.data.target.classification}, area {results['area']}, entropy {results['entropy']},",  # type: ignore
                f"insertion curve {results['insertion_curve']}, deletion curve {results['deletion_curve']}",
            )

    end = time.time()
    time_taken = end - start
    logger.info(f"Time taken: {time_taken:.2f}s")

    if args.surface is not None:
        if path is not None:
            pass
        elif args.surface == "show":
            path = None
        else:
            path = args.surface
        exp.surface_plot(path)

    if args.heatmap is not None:
        if args.heatmap == "show":
            path = None
        else:
            path = args.heatmap
        exp.heatmap_plot(path)

    if args.output is not None:
        if path is None:
            if args.output == "show":
                path = None
            else:
                path = args.output
        exp.save(path, clauses=clauses)

    if db is not None:
        if args.strategy == Strategy.MultiSpotlight:
            logger.info("writing multiple explanations to database")
            update_database(
                db,
                exp,
                time_taken,
                multi=True,
            )
        else:
            logger.info("writing to database")
            update_database(
                db,
                exp,
                time_taken,
            )

    return exp


def get_prediction_func_from_args(args: CausalArgs):
    """Takes a CausalArgs object and gets the prediction function and model shape.

    If ``args.script`` specifies a prediction function and model shape, returns these.
    Otherwise gets the prediction function and model shape from the provided model
    file.

    Args:
        args: configuration values for ReX

    Returns:
        tuple containing

        - ``prediction_func``
        - ``model_shape``

    Raises:
        RuntimeError: if an onnx inference instance cannot be created from the provided model file.

    """
    if hasattr(args.script, "prediction_function") and hasattr(
        args.script, "model_shape"
    ):
        prediction_func = args.script.prediction_function  # type: ignore
        model_shape = args.script.model_shape()  # type: ignore
    else:
        ps = get_prediction_function(args)
        if ps is None:
            raise RuntimeError("Unable to create an onnx inference instance")
        else:
            prediction_func, model_shape = ps

    return prediction_func, model_shape


def explanation(
    args: CausalArgs, device: tt.device, db: Session | None = None
) -> Union[Explanation, List[Explanation]]:
    """Takes a CausalArgs object and returns a Explanation.

    Takes a CausalArgs object and returns either an Explanation, or a list of Explanations
    if the input ``args.path`` is a directory rather than a path to a single file.

    Args:
        args: configuration values for ReX
        device: as returned by :py:func:`~rex_xai._utils.get_device()`
        db: None or as returned by :py:func:`~rex_xai.database.initialise_rex_db()`

    Returns:
        Explanation:
            An :py:class:`~rex_xai.extraction.Explanation` object containing the causal reponsibility explanation
            calculated using the given ``args``.

    """

    prediction_func, model_shape = get_prediction_func_from_args(args)

    if isinstance(model_shape[0], int) and model_shape[0] < args.batch_size:
        logger.warning(
            f"Resetting batch size to size of model's first axis: {model_shape[0]}"
        )
        args.batch_size = model_shape[0]

    # directory of data to process
    if os.path.isdir(args.path):
        explanations = []
        dir = args.path
        path = None
        for dir, _, files in os.walk(args.path):
            for f in files:
                to_process = os.path.join(dir, f)
                logger.info("processing %s", to_process)
                current_args = copy.copy(args)
                current_args.path = to_process
                if args.output is not None and args.output != "show":
                    out_dir = os.path.dirname(args.output)
                    name, ext = os.path.splitext(args.output)
                    fname, _ = os.path.splitext(f)
                    path = f"{out_dir}_{fname}_{name}{ext}"
                exp = _explanation(
                    current_args,
                    model_shape,
                    prediction_func,
                    device,
                    db,
                    path=path,
                )
                explanations.append(exp)
        return explanations

    else:
        # a single explanation
        return _explanation(args, model_shape, prediction_func, device, db)
