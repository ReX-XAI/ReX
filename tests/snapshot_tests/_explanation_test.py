import pytest
from rex_xai.explanation import _explanation, analyze
from rex_xai._utils import Strategy

@pytest.mark.parametrize("batch_size", [1, 64])
def test__explanation_snapshot(
    args_custom, model_shape, prediction_func, cpu_device, batch_size, snapshot_explanation
):
    args_custom.batch_size = batch_size
    exp = _explanation(args_custom, model_shape, prediction_func, cpu_device, db=None)

    assert exp == snapshot_explanation
    assert hash(tuple(exp.explanation.reshape(-1).tolist())) == snapshot_explanation


@pytest.mark.parametrize("batch_size", [1, 64])
def test__explanation_snapshot_diff_model_shape(
    args_torch_swin_v2_t,
    model_shape_swin_v2_t,
    prediction_func_swin_v2_t,
    cpu_device,
    batch_size,
    snapshot_explanation
):
    args_torch_swin_v2_t.batch_size = batch_size

    exp = _explanation(
        args_torch_swin_v2_t,
        model_shape_swin_v2_t,
        prediction_func_swin_v2_t,
        cpu_device,
        db=None,
    )

    assert exp == snapshot_explanation
    assert hash(tuple(exp.explanation.reshape(-1).tolist())) == snapshot_explanation


@pytest.mark.parametrize("strategy", [Strategy.Global, Strategy.Spatial])
def test_extract_analyze(exp_custom, strategy, snapshot):
    exp_custom.extract(strategy)
    results = analyze(exp_custom, "RGB")
    results_rounded = {k: round(v, 4) for k, v in results.items() if v is not None}

    assert hash(tuple(exp_custom.final_mask.reshape(-1).tolist())) == snapshot
    assert results_rounded == snapshot
