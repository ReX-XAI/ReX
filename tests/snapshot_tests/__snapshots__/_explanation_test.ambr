# serializer version: 1
# name: test__explanation_snapshot[1]
  Explanation(
    args=CausalArgs(
      all=False,
      alpha=0.2,
      analyze=False,
      batch_size=1,
      binary_threshold=None,
      blend=0.0,
      chunk_size=10,
      colour=200,
      concentrate=False,
      confidence_filter=0.0,
      config_location=None,
      data_location=None,
      db=None,
      distribution=<Distribution.Uniform: 2>,
      distribution_args=None,
      gpu=False,
      grid=False,
      heatmap=None,
      heatmap_colours='magma',
      info=True,
      insertion_step=100,
      iters=2,
      mark_segments=False,
      mask_value=0,
      max_spotlight_budget=40,
      means=None,
      min_box_size=10,
      minimum_confidence_threshold=0.0,
      mode=None,
      multi_style='composite',
      no_expansions=4,
      norm=255.0,
      normalise_curves=True,
      output=None,
      path='tests/test_data/dog.jpg',
      permitted_overlap=0.0,
      processed=False,
      progress_bar=True,
      queue_len=1,
      queue_style=<Queue.Area: 1>,
      raw=False,
      resize=False,
      search_limit=1000,
      seed=42,
      segmentation=False,
      shape=None,
      spatial_initial_radius=25,
      spatial_radius_eta=0.2,
      spotlight_eta=0.2,
      spotlight_size=20,
      spotlight_step=5,
      spotlights=10,
      stds=None,
      strategy=<Strategy.Global: 1>,
      surface=None,
      tree_depth=10,
      verbosity=0,
      weighted=False,
    ),
    data=Data: RGB, ['N', 3, 224, 224], 224, 224, 3, first
  	 Target:FOUND_CLASS: 207, FOUND_CONF: 0.20462, TARGET_CLASS: n/a, TARGET_CONFIDENCE: n/a,
    explanation_confidence=0.015131033025681973,
    maps={207: 1},
    run_stats=dict({
      'avg_box_size': 183.5625,
      'max_depth_reached': 5,
      'total_failing': 126,
      'total_passing': 84,
    }),
  )
# ---
# name: test__explanation_snapshot[1].1
  2537636458340732753
# ---
# name: test__explanation_snapshot[64]
  Explanation(
    args=CausalArgs(
      all=False,
      alpha=0.2,
      analyze=False,
      batch_size=64,
      binary_threshold=None,
      blend=0.0,
      chunk_size=10,
      colour=200,
      concentrate=False,
      confidence_filter=0.0,
      config_location=None,
      data_location=None,
      db=None,
      distribution=<Distribution.Uniform: 2>,
      distribution_args=None,
      gpu=False,
      grid=False,
      heatmap=None,
      heatmap_colours='magma',
      info=True,
      insertion_step=100,
      iters=2,
      mark_segments=False,
      mask_value=0,
      max_spotlight_budget=40,
      means=None,
      min_box_size=10,
      minimum_confidence_threshold=0.0,
      mode=None,
      multi_style='composite',
      no_expansions=4,
      norm=255.0,
      normalise_curves=True,
      output=None,
      path='tests/test_data/dog.jpg',
      permitted_overlap=0.0,
      processed=False,
      progress_bar=True,
      queue_len=1,
      queue_style=<Queue.Area: 1>,
      raw=False,
      resize=False,
      search_limit=1000,
      seed=42,
      segmentation=False,
      shape=None,
      spatial_initial_radius=25,
      spatial_radius_eta=0.2,
      spotlight_eta=0.2,
      spotlight_size=20,
      spotlight_step=5,
      spotlights=10,
      stds=None,
      strategy=<Strategy.Global: 1>,
      surface=None,
      tree_depth=10,
      verbosity=0,
      weighted=False,
    ),
    data=Data: RGB, ['N', 3, 224, 224], 224, 224, 3, first
  	 Target:FOUND_CLASS: 207, FOUND_CONF: 0.20462, TARGET_CLASS: n/a, TARGET_CONFIDENCE: n/a,
    explanation_confidence=0.01513106282800436,
    maps={207: 1},
    run_stats=dict({
      'avg_box_size': 183.5625,
      'max_depth_reached': 5,
      'total_failing': 126,
      'total_passing': 84,
    }),
  )
# ---
# name: test__explanation_snapshot[64].1
  2537636458340732753
# ---
# name: test__explanation_snapshot_diff_model_shape[1]
  Explanation(
    args=CausalArgs(
      all=False,
      alpha=0.2,
      analyze=False,
      batch_size=1,
      binary_threshold=None,
      blend=0.0,
      chunk_size=10,
      colour=200,
      concentrate=False,
      confidence_filter=0.0,
      config_location=None,
      data_location=None,
      db=None,
      distribution=<Distribution.Uniform: 2>,
      distribution_args=None,
      gpu=False,
      grid=False,
      heatmap=None,
      heatmap_colours='magma',
      info=True,
      insertion_step=100,
      iters=2,
      mark_segments=False,
      mask_value=0,
      max_spotlight_budget=40,
      means=None,
      min_box_size=10,
      minimum_confidence_threshold=0.0,
      mode=None,
      multi_style='composite',
      no_expansions=4,
      norm=255.0,
      normalise_curves=True,
      output=None,
      path='tests/test_data/dog.jpg',
      permitted_overlap=0.0,
      processed=False,
      progress_bar=True,
      queue_len=1,
      queue_style=<Queue.Area: 1>,
      raw=False,
      resize=False,
      search_limit=1000,
      seed=42,
      segmentation=False,
      shape=None,
      spatial_initial_radius=25,
      spatial_radius_eta=0.2,
      spotlight_eta=0.2,
      spotlight_size=20,
      spotlight_step=5,
      spotlights=10,
      stds=None,
      strategy=<Strategy.Global: 1>,
      surface=None,
      tree_depth=10,
      verbosity=0,
      weighted=False,
    ),
    data=Data: RGB, ['N', 3, 256, 256], 256, 256, 3, first
  	 Target:FOUND_CLASS: 215, FOUND_CONF: 0.48740, TARGET_CLASS: n/a, TARGET_CONFIDENCE: n/a,
    explanation_confidence=0.037296317517757416,
    maps={215: 1},
    run_stats=dict({
      'avg_box_size': 262.0,
      'max_depth_reached': 5,
      'total_failing': 136,
      'total_passing': 60,
    }),
  )
# ---
# name: test__explanation_snapshot_diff_model_shape[1].1
  -5471936325618413956
# ---
# name: test__explanation_snapshot_diff_model_shape[64]
  Explanation(
    args=CausalArgs(
      all=False,
      alpha=0.2,
      analyze=False,
      batch_size=64,
      binary_threshold=None,
      blend=0.0,
      chunk_size=10,
      colour=200,
      concentrate=False,
      confidence_filter=0.0,
      config_location=None,
      data_location=None,
      db=None,
      distribution=<Distribution.Uniform: 2>,
      distribution_args=None,
      gpu=False,
      grid=False,
      heatmap=None,
      heatmap_colours='magma',
      info=True,
      insertion_step=100,
      iters=2,
      mark_segments=False,
      mask_value=0,
      max_spotlight_budget=40,
      means=None,
      min_box_size=10,
      minimum_confidence_threshold=0.0,
      mode=None,
      multi_style='composite',
      no_expansions=4,
      norm=255.0,
      normalise_curves=True,
      output=None,
      path='tests/test_data/dog.jpg',
      permitted_overlap=0.0,
      processed=False,
      progress_bar=True,
      queue_len=1,
      queue_style=<Queue.Area: 1>,
      raw=False,
      resize=False,
      search_limit=1000,
      seed=42,
      segmentation=False,
      shape=None,
      spatial_initial_radius=25,
      spatial_radius_eta=0.2,
      spotlight_eta=0.2,
      spotlight_size=20,
      spotlight_step=5,
      spotlights=10,
      stds=None,
      strategy=<Strategy.Global: 1>,
      surface=None,
      tree_depth=10,
      verbosity=0,
      weighted=False,
    ),
    data=Data: RGB, ['N', 3, 256, 256], 256, 256, 3, first
  	 Target:FOUND_CLASS: 215, FOUND_CONF: 0.48740, TARGET_CLASS: n/a, TARGET_CONFIDENCE: n/a,
    explanation_confidence=0.0372963547706604,
    maps={215: 1},
    run_stats=dict({
      'avg_box_size': 262.0,
      'max_depth_reached': 5,
      'total_failing': 136,
      'total_passing': 60,
    }),
  )
# ---
# name: test__explanation_snapshot_diff_model_shape[64].1
  -5471936325618413956
# ---
# name: test_extract_analyze[Strategy.Global]
  -2185390445883109608
# ---
# name: test_extract_analyze[Strategy.Global].1
  dict({
    'area': 0.0037,
    'deletion_curve': 1.0199,
    'entropy': 7.1353,
    'insertion_curve': 1.2399,
  })
# ---
# name: test_extract_analyze[Strategy.Spatial]
  -2185390445883109608
# ---
# name: test_extract_analyze[Strategy.Spatial].1
  dict({
    'area': 0.0037,
    'deletion_curve': 1.0199,
    'entropy': 7.1353,
    'insertion_curve': 1.2399,
  })
# ---
